src/crimson/common/errorator.h

```diff
@@ -0,0 +1,827 @@
+  };
+
+  template <typename... T, typename... A>
+  static future<T...> make_ready_future(A&&... value) {
+    return future<T...>(ready_future_marker(), std::forward<A>(value)...);
+  }
+
+  static auto now() {
+    return make_ready_future<>();
+  }
+
+private:
+  template <class... Args>
+  static decltype(auto) plainify(seastar::future<Args...>&& fut) {
+    return std::forward<seastar::future<Args...>>(fut);
+  }
+  template <class Arg>
+  static decltype(auto) plainify(Arg&& arg) {
+    return std::forward<typename Arg::base_t>(arg);
+  }
+
+  template <class T, class = std::void_t<T>>
+  class futurize {
+    using vanilla_futurize = seastar::futurize<T>;
+
+    template <class...>
+    struct tuple2future {};
+    template <class... Args>
+    struct tuple2future <std::tuple<Args...>> {
+      using type = future<Args...>;
+    };
+
+  public:
+    using type =
+      typename tuple2future<typename vanilla_futurize::value_type>::type;
+
+    template <class Func, class... Args>
+    static type apply(Func&& func, std::tuple<Args...>&& args) {
+      return vanilla_futurize::apply(std::forward<Func>(func),
+                                     std::forward<std::tuple<Args...>>(args));
+    }
+
+    template <typename Arg>
+    static type make_exception_future(Arg&& arg) {
+      return vanilla_futurize::make_exception_future(std::forward<Arg>(arg));
+    }
+  };
+  template <template <class...> class ErroratedFutureT,
+            class... ValuesT>
+  class futurize<ErroratedFutureT<::ceph::errorated_future_marker<ValuesT...>>,
+                 std::void_t<
+                   typename ErroratedFutureT<
+                     ::ceph::errorated_future_marker<ValuesT...>>::errorator_type>> {
+  public:
+    using type = ::ceph::errorator<AllowedErrors...>::future<ValuesT...>;
+
+    template <class Func, class... Args>
+    static type apply(Func&& func, std::tuple<Args...>&& args) {
+      try {
+        return ::seastar::apply(std::forward<Func>(func),
+                                std::forward<std::tuple<Args...>>(args));
+      } catch (...) {
+        return make_exception_future(std::current_exception());
+      }
+    }
+
+    template <typename Arg>
+    static type make_exception_future(Arg&& arg) {
+      return ::seastar::make_exception_future<ValuesT...>(std::forward<Arg>(arg));
+    }
+  };
+
+  template <class ErrorT>
+  static std::exception_ptr make_exception_ptr(ErrorT&& e) {
+    // calling via interface class due to encapsulation and friend relations.
+    return e.error_t<std::decay_t<ErrorT>>::to_exception_ptr();
+  }
+
+  // needed because of:
+  //  * ReturnErrorator::template futurize<...> in `safe_then()`,
+  //  * conversion to `std::exception_ptr` in `future::future(ErrorT&&)`.
+  // the friendship with all errorators is an idea from Kefu to fix build
+  // issues on GCC 9. This version likely fixes some access violation bug
+  // we were exploiting before.
+  template <class>
+  friend class _future;
```

Comment:
this friend is not necessary. as members always have the full access of the enclosing class, except for `this`.

Reply:
Addressed.